### DISCUSSION TITLE
Can you call `set_fills()` more than once on a Py5Shape object?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#577] [Q&A]
Season's greetings!

![image](https://github.com/user-attachments/assets/3fffe993-5675-41c6-94fc-ad10184251a2)

Am I doing something wrong? It looks like something happens after a call to `.set_fills()`. Maybe Py5Shape objects get tired after a long year? :smile: 

Maybe something fun to investigate in 2025!

```python
import py5
import numpy as np

def setup():
    global icosa
    py5.size(600, 600, py5.P3D)
    icosa = icosahedron(250)    
    icosa.set_fills(random_color() for _ in range(60))

def draw():
    py5.background(0)
#     if py5.frame_count > 100:
#         icosa.set_fills(random_color() for _ in range(60))
    py5.translate(py5.width / 2, py5.height / 2)
    py5.shape(icosa, 0, 0)

def random_color():
    return py5.color(py5.random_int(255), py5.random_int(255), py5.random_int(255))

def key_pressed():
    print(py5.key)
    if py5.key == 'a':
        icosa.set_fills(random_color() for _ in range(60))
    elif py5.key == 'b':
        icosa.set_fill(random_color())

def icosahedron(radius):
    global vs, sequence
    f = (radius / 2) # / 0.95 # aprox. ?
    phi = (py5.sqrt(5) + 1) / 2
    a = phi * f
    b = 1 * f
    vs = np.array([
        (a,  b,  0),
        (a,  -b,  0),
        (-a,  -b,  0),
        (-a,  b,  0),
        (b,  0, a),
        (-b,  0, a),
        (-b,  0, -a),
        (b, 0, -a),
        (0, a, b),
        (0, a, -b),
        (0, -a, -b),
        (0, -a, b),
    ])
    sequence = np.array([
        2, 11, 5, 2, 5, 3, 5, 8, 3, 4, 8, 5, 11, 4, 5,
        10, 2, 6, 2, 3, 6, 3, 9, 6, 3, 8, 9, 2, 10, 11,
        1, 4, 11, 1, 0, 4, 7, 0, 1, 10, 1, 11, 0, 8, 4,
        0, 9, 8, 7, 9, 0, 1, 10, 7, 10, 6, 7, 6, 9, 7,  
    ])
    shp = py5.create_shape()
    shp.begin_shape(py5.TRIANGLES)
    shp.vertices(vs[sequence])
    shp.end_shape()
    return shp

py5.run_sketch(block=False)
```

**Comment by vsquared** [comment]
You probably already know this, but the following works on my mac.  To get it to run I had to add a _def settings():_ to setup the window and set blocking to _True_.  For some reason **icosa** is not recognized by key_pressed(): as being a global which I don't understand since draw(): seems to recognize it.  By defining **icosa** a second time inside of key_pressed() it seems to fix it:
```python
import py5
import numpy as np

def settings():
    py5.size(600, 600, py5.P3D)
    
def setup():
    global icosa
    icosa = icosahedron(250)    
    icosa.set_fills(random_color() for _ in range(60))

def draw():
    py5.background(0)
    py5.translate(py5.width / 2, py5.height / 2)
    py5.shape(icosa, 0, 0)

def random_color():
    return py5.color(py5.random_int(255), py5.random_int(255), py5.random_int(255))

def key_pressed():
    global icosa;
    print(py5.key)
    if py5.key == 'a':
        icosa = icosahedron(250) 
        icosa.set_fills(random_color() for _ in range(60))
    elif py5.key == 'b':
        icosa.set_fill(random_color())

def icosahedron(radius):
    global vs, sequence
    f = (radius / 2) # / 0.95 # aprox. ?
    phi = (py5.sqrt(5) + 1) / 2
    a = phi * f
    b = 1 * f
    vs = np.array([
        (a,  b,  0),
        (a,  -b,  0),
        (-a,  -b,  0),
        (-a,  b,  0),
        (b,  0, a),
        (-b,  0, a),
        (-b,  0, -a),
        (b, 0, -a),
        (0, a, b),
        (0, a, -b),
        (0, -a, -b),
        (0, -a, b),
    ])
    sequence = np.array([
        2, 11, 5, 2, 5, 3, 5, 8, 3, 4, 8, 5, 11, 4, 5,
        10, 2, 6, 2, 3, 6, 3, 9, 6, 3, 8, 9, 2, 10, 11,
        1, 4, 11, 1, 0, 4, 7, 0, 1, 10, 1, 11, 0, 8, 4,
        0, 9, 8, 7, 9, 0, 1, 10, 7, 10, 6, 7, 6, 9, 7,  
    ])
    shp = py5.create_shape()
    shp.begin_shape(py5.TRIANGLES)
    shp.vertices(vs[sequence])
    shp.end_shape()
    return shp

py5.run_sketch(block=True)
```

**Reply by villares** [reply]
Cheers @vsquared! 

On my [yesterday's sketch-a-day](https://github.com/villares/sketch-a-day/tree/main/2024/sketch_2024_12_25) I did this as a workaround... recreating the Py5Shape object on each frame :)

You needed the `global` statement in `key_pressed()` because you were doing a new assignment on the global name `icosa`, in `draw()` you are just "reading" the variable, so there is no need to add `global`. This was one of the trickiest things for me coming from Java to Python.

On MacOS the conventional Python interpreter won't let you set `block=False` :(

**Comment by hx2A** [comment]
I just tried running this. I believe it might be a Processing bug where once the shape is drawn to the screen, Processing's PShapeOpenGL method `setFill(int index, int fill)` (which is what py5 calls in a loop when you call Py5Shape's `set_fills()` method) does not let you update the shape. Can you try testing it to see? From looking at Processing's source, I see that Processing's `setFill()` and `setFills()` behave a bit differently if the shape has been tesselated (this happens when it is rendered to the screen for the first time).

**Reply by villares** [reply]
![image](https://github.com/user-attachments/assets/d64affa4-b17a-45d9-b57d-2c73be21d576)

Is `setFills()` undocumented or "internal only" in Processing? I could only find `setFill()` https://processing.org/reference/PShape.html :confused: 

`setFill()` (and equivalent `set_fill()` in py5) can be called and work as expected (making all fill the same).

```pde
PShape shp;

void setup(){
  size(300, 300, P3D);
  shp = createShape();
  shp.beginShape(TRIANGLES);
  shp.vertex(100, 100);
  shp.vertex(200, 100);
  shp.vertex(100, 200);
  shp.vertex(200, 200);
  shp.vertex(200, 100);
  shp.vertex(100, 200);
  shp.endShape();
}

void draw(){
  shp.setFill(color(random(255), 200, 0));
  shape(shp);
}
```

![peek_1](https://github.com/user-attachments/assets/2634f419-3cf6-48a9-9875-70171f93e793)

**Reply by hx2A** [reply]
Sorry, `setFills()` is a py5 addition. Processing doesn't have it.

The source for `setFills()` is basically a for loop:

```java
      for (int i = 0; i < fills.length; ++i) {
        s.setFill(i, fills[i]);
      }
```

So to test it, you'd basically need compare `PShape.setfill(color)` with `PShape.setfill(i, color)` where `i` is a for loop.

**Comment by vsquared** [comment]
The following Processing code demonstrates a PShape array for the sides of a cube and then using a 'for' loop to change each side's color with the setFills() code shown above.

```python
// Reference for cube:
//https://stackoverflow.com/questions/72270464/how-to-draw-a-cube-in-processing-without-the-box-function

PShape cube;
PShape[] sides = new PShape[6];
color[] fills = new color[6];

void setup() {
  size(600, 600, P3D);
  surface.setTitle("Press mouse to change colors.");
  strokeWeight(3);
  for (int i = 0; i < fills.length; ++i) {
    fills[i] = color(random(255), random(255), random(255));
  }
  cube = createShape(GROUP);
  cubeSides(150);
}

void draw() {
  background(255);
  translate(width/2, height/2, 0);
  rotateY(map(mouseX, 0, width, -PI, PI));
  rotateX(map(mouseY, 0, height, PI, -PI));
  shape(cube);
}

void cubeSides(float size) {
  // half size: keep the pivot at the center
  float s = size * 0.5;
  // back (-z)
  sides[0] = createShape();
  sides[0].beginShape();
  sides[0].fill(fills[0]);
  sides[0].vertex(-s, -s, -s);
  sides[0].vertex(+s, -s, -s);
  sides[0].vertex(+s, +s, -s);
  sides[0].vertex(-s, +s, -s);
  sides[0].endShape();
  // front (+z)
  sides[1] = createShape();
  sides[1].beginShape();
  sides[1].fill(fills[1]);
  sides[1].vertex(-s, -s, +s);
  sides[1].vertex(+s, -s, +s);
  sides[1].vertex(+s, +s, +s);
  sides[1].vertex(-s, +s, +s);
  sides[1].endShape();
  // top (-y)
  sides[2] = createShape();
  sides[2].beginShape();
  sides[2].fill(fills[2]);
  sides[2].vertex(-s, -s, +s);
  sides[2].vertex(-s, -s, -s);
  sides[2].vertex(+s, -s, -s);
  sides[2].vertex(+s, -s, +s);
  sides[2].endShape();
  // bottom (+y)
  sides[3] = createShape();
  sides[3].beginShape();
  sides[3].fill(fills[3]);
  sides[3].vertex(+s, +s, +s);
  sides[3].vertex(+s, +s, -s);
  sides[3].vertex(-s, +s, -s);
  sides[3].vertex(-s, +s, +s);
  sides[3].endShape();
  // left (-x)
  sides[4] = createShape();
  sides[4].beginShape();
  sides[4].fill(fills[4]);
  sides[4].vertex(-s, -s, +s);
  sides[4].vertex(-s, -s, -s);
  sides[4].vertex(-s, +s, -s);
  sides[4].vertex(-s, +s, +s);
  sides[4].endShape();
  // right (+x)
  sides[5] = createShape();
  sides[5].beginShape();
  sides[5].fill(fills[5]);
  sides[5].vertex(+s, -s, +s);
  sides[5].vertex(+s, -s, -s);
  sides[5].vertex(+s, +s, -s);
  sides[5].vertex(+s, +s, +s);
  sides[5].endShape();
  
  cube.addChild(sides[0]);
  cube.addChild(sides[1]);
  cube.addChild(sides[2]);
  cube.addChild(sides[3]);
  cube.addChild(sides[4]);
  cube.addChild(sides[5]);
}

void mousePressed() {
  // Py5 set_fills() code:
  for (int i = 0; i < fills.length; ++i) {
    fills[i] = color(random(255), random(255), random(255));
    sides[i].setFill(fills[i]);
  }
}

```
**Py5 version:**
Uses set_fill() in a 'for' loop.  set_fills() apparently requires numpy which I am not familiar with. Requires Imported mode for py5.
```python
sides = []
fills = []

def setup():
  global cube;
  size(600, 600, P3D)
  window_title("Press mouse to change colors.")
  stroke_weight(3)
  for i in range(6):
      fills.append(color(random(255),random(255),random(255)))
  cube = create_shape(GROUP)
  cubeSides(150)

def draw():
  background(255)
  translate(width/2, height/2, 0)
  rotate_y(remap(mouse_x, 0, width, -PI, PI))
  rotate_x(remap(mouse_y, 0, height, PI, -PI))
  shape(cube)

def cubeSides(size):
  global sides
  global fills
  global cube
#    half size: keep the pivot at the center
  s = size * 0.5;
  for i in range(6):
    sides.append(create_shape())
#    back (-z)
  sides[0].begin_shape()
  sides[0].fill(fills[0])
  sides[0].vertex(-s, -s, -s)
  sides[0].vertex(+s, -s, -s)
  sides[0].vertex(+s, +s, -s)
  sides[0].vertex(-s, +s, -s)
  sides[0].end_shape()
#    front (+z)
  sides[1].begin_shape()
  sides[1].fill(fills[1])
  sides[1].vertex(-s, -s, +s)
  sides[1].vertex(+s, -s, +s)
  sides[1].vertex(+s, +s, +s)
  sides[1].vertex(-s, +s, +s)
  sides[1].end_shape()
#    top (-y)
  sides[2].begin_shape();
  sides[2].fill(fills[2])
  sides[2].vertex(-s, -s, +s)
  sides[2].vertex(-s, -s, -s)
  sides[2].vertex(+s, -s, -s)
  sides[2].vertex(+s, -s, +s)
  sides[2].end_shape()
#    bottom (+y)
  sides[3].begin_shape()
  sides[3].fill(fills[3])
  sides[3].vertex(+s, +s, +s)
  sides[3].vertex(+s, +s, -s)
  sides[3].vertex(-s, +s, -s)
  sides[3].vertex(-s, +s, +s)
  sides[3].end_shape()
#    left (-x)
  sides[4].begin_shape()
  sides[4].fill(fills[4])
  sides[4].vertex(-s, -s, +s)
  sides[4].vertex(-s, -s, -s)
  sides[4].vertex(-s, +s, -s)
  sides[4].vertex(-s, +s, +s)
  sides[4].end_shape()
#    right (+x)
  sides[5].begin_shape()
  sides[5].fill(fills[5])
  sides[5].vertex(+s, -s, +s)
  sides[5].vertex(+s, -s, -s)
  sides[5].vertex(+s, +s, -s)
  sides[5].vertex(+s, +s, +s)
  sides[5].end_shape()

  cube.add_child(sides[0])
  cube.add_child(sides[1])
  cube.add_child(sides[2])
  cube.add_child(sides[3])
  cube.add_child(sides[4])
  cube.add_child(sides[5])

def mouse_pressed():
  for i in range(6):
      sides[i].set_fill(color(random_int(255), random_int(255), random_int(255)))

```

<img width="482" alt="sfs" src="https://github.com/user-attachments/assets/babda0d4-d17c-4bac-9f69-55999022c61e" />

**Reply by hx2A** [reply]
Good thinking! This is a good approach that allows you to change each side's color.

**Reply by villares** [reply]
Thanks for this @vsquared! I'll try to use your code to test the issue! 

You can use numpy on imported mode, just add `import numpy as np` to your sketch. In 2025 I'd like to write more introductory materials about numpy, I'm kind of a beginner myself and I'm inspired by people like [Tariq Hashid](https://www.youtube.com/@AlgorithmicArt) who can write teaching stuff as they learn a new subject themselves.

**Reply by vsquared** [reply]
I can import it, I just don't know how to use it; looks complicated to me.  I don't understand why it appears to be required:
<img width="470" alt="set_fills_err" src="https://github.com/user-attachments/assets/273b18c6-5efd-448d-90eb-7b373cf31a1b" />

**Reply by villares** [reply]
I wasn't aware it was required :confused: ! A numpy array can be a bit like a Python list but maybe more like a Java array (fixed size and only one type of value allowed, they call it homogeneous). Try this:
```python
# a list comprehension
list_of_colors = [color(random(255), random(255), random(255)) for i in range(number_of_colors)]
# converting a list to a NumPy array
array_of_colors = np.array(list_of_colors)
# then try
shp.set_fills(array_of_colors)
```

**Reply by hx2A** [reply]
> I can import it, I just don't know how to use it; looks complicated to me. I don't understand why it appears to be required:

It could be a numpy array, and a numpy array is fastest. However, you can use a regular Python list or (I believe) a generator. It will convert those other things to an array for you.

**Reply by villares** [reply]
> It could be a numpy array, and a numpy array is fastest. However, you can use a regular Python list or (I believe) a generator. It will convert those other things to an array for you.

Yeah, I remember having this discussion so we could have [generator expressions](https://peps.python.org/pep-0289/), that's what I thought, but [the signature in the documentation](https://py5coding.org/reference/py5shape_set_fills.html) is not showing that... it could be a doc bug or something else (I should check this later).

**Reply by hx2A** [reply]
There might be a way to add it, but I don't want the signature to look like

```set_fill(fills: Union[list, npt.NDArray[np.int32]], /) -> None```

or something even more horrific. Also, adding it would mean adding everywhere, which might be an undertaking.

**Reply by villares** [reply]
I see @hx2A, no worries. And looking closely at the example in the docs, it is done with a list!

**Reply by hx2A** [reply]
Let me think about it some more. I opened an issue #584 so I remember to come back to this and see if there is a way to improve it.

**Reply by villares** [reply]
Can the "customized errors" infrastructure be used to help somehow?

**Reply by vsquared** [reply]
I question whether set_fills() will work on a three dimensional object.  The example code is for a two dimensional rectangle and it works as expected where the number of vertices is equal to the number of sides.  The original example of this thread is trying to change the face colors of a multifaceted 3D shape which may not have an equal number of vertices.  If the number of faces does not equal the number of vertices, the following error is generated when trying to use set_fills on a PShape cube (six faces, four vertices per shape):
<img width="923" alt="e" src="https://github.com/user-attachments/assets/75a68ef5-e27f-4ef1-8dcf-34694f2379b0" />

**Reply by hx2A** [reply]
No, `set_fills()` should work fine on a 3D object. There must be a bug where it doesn't work after the shape is drawn to the screen for the first time.

`set_fills()` won't work on a group shape though. You'd need to call it for each of the child shapes and have the correct number of fills, which will equal the number of vertices. Each vertex has its own fill color.

**Reply by vsquared** [reply]
I tried that and couldn't get it to work (but it doesn't crash):
```python
def mouse_pressed():
  for child in cube.get_children():
      child.set_fills(color(random(255),random(255),random(255)) for _ in range(4))
```
If I use range(6) which is the number of faces, I get this:
<img width="933" alt="e2" src="https://github.com/user-attachments/assets/6b812892-79ca-46c7-b6c7-571e96844b3b" />

**Reply by hx2A** [reply]
> I tried that and couldn't get it to work (but it doesn't crash):

Right, it doesn't work. That's the bug. After the shape is drawn to the screen, `set_fills()` can no longer change the vertex fills.

You could use `set_fill()` and set the child object to a new (single) fill color. That works.

> If I use range(6) which is the number of faces, I get this:

You should use `range(4)` there because each child shape (each face of the cube) has 4 vertices. Six is the number of cube faces or the number of children in `cube`.

**Reply by vsquared** [reply]
`You could use set_fill() and set the child object to a new (single) fill color. That works.`
Correct.  That's why I don't think you need set_fills(), but that's just my opinion.  If someone can produce a workable 3D multifaceted demo I could change my mind.  Even that apparently precludes iterating solid colors for the faces which would still require set_fill() in a 'for' loop.
**Question:**
If set_fills() is calling set_fill() in a 'for' loop behind the curtain and set_fill() with a 'for' loop works by itself then why doesn't set_fills() work as expected?  Things must not be what they seem behind the curtain.

**Reply by hx2A** [reply]
> Correct. That's why I don't think you need set_fills(), but that's just my opinion.

Right, it isn't useful for this case. It isn't useful for a lot of things outside of what it was created for, which is helping py5 convert Trimesh objects to Py5Shape objects. And that it does very well.

> If set_fills() is calling set_fill() in a 'for' loop behind the curtain and set_fill() with a 'for' loop works by itself then why doesn't set_fills() work as expected?

py5's `set_fills()` calls Processing's `setFill(int index, int color)` in a for loop, which has a different implementation than Processing's `setFill(int color)`.  Processing's `setFill(int color)` does not call Processing's `setFill(int index, int color)` in a for loop. The differences are an important clue about why it isn't working.

**Comment by vsquared** [comment]
**Alternate way of using setFills() more than once:**
```python
# Uses Imported mode for py5

def drawMyCube():
  global cube
  cube = create_shape()
  s = 75
  cube.begin_shape(QUADS)
# back (-z)
  cube.vertex(-s, -s, -s)
  cube.vertex(+s, -s, -s)
  cube.vertex(+s, +s, -s)
  cube.vertex(-s, +s, -s)
# front (+z)  
  cube.vertex(-s, -s, +s)
  cube.vertex(+s, -s, +s)
  cube.vertex(+s, +s, +s)
  cube.vertex(-s, +s, +s)
# top (-y)
  cube.vertex(-s, -s, +s)
  cube.vertex(-s, -s, -s)
  cube.vertex(+s, -s, -s)
  cube.vertex(+s, -s, +s)
# bottom (+y)
  cube.vertex(+s, +s, +s)
  cube.vertex(+s, +s, -s)
  cube.vertex(-s, +s, -s)
  cube.vertex(-s, +s, +s)
# left (-x)
  cube.vertex(-s, -s, +s)
  cube.vertex(-s, -s, -s)
  cube.vertex(-s, +s, -s)
  cube.vertex(-s, +s, +s)
# right (+x)
  cube.vertex(+s, -s, +s)
  cube.vertex(+s, -s, -s)
  cube.vertex(+s, +s, -s)
  cube.vertex(+s, +s, +s)
  cube.end_shape()
  cube.set_fills(color(random(255),random(255),random(255)) for _ in range(24))

def setup():
  size(600, 600, P3D)
  window_title("Rotate with mouse click to change colors.")
  stroke_weight(3)
  drawMyCube()

def draw():
  background(209)
  translate(width/2, height/2, 0)
  rotate_x(remap(mouse_y, 0, height, PI, -PI))
  rotate_y(remap(mouse_x, 0, width, -PI, PI))
  shape(cube)

def mouse_pressed():
  drawMyCube()  

```
**Output:**
<img width="316" alt="set_fills" src="https://github.com/user-attachments/assets/f49c397f-3f66-4b70-b937-925968ad3880" />

**Reply by hx2A** [reply]
This looks great, @vsquared 

`set_fills()` works in here because it is called before the first time the cube is drawn to the screen. The problem @villares was having is that he tried to call `set_fills()` after the object is drawn to the screen. In Processing, the fill colors seem to become permanent once the object is drawn for the first time.

**Reply by vsquared** [reply]
`set_fills() works in here because it is called before the first time the cube is drawn to the screen.`
Doesn't calling it from mousePressed() qualify as being called a second time?

**Reply by villares** [reply]
The thing is a new fresh object is being generated each time you call DrawMyCube, by the `cube = create_shape()` and subsequent instructions. So each time `set_fills()` is acting only once on these new objects.

I should at some point try more experiments in Processing Java to investigate this, because, if it worked, it would open out some performance gain opportunities, since reconstructing the objects every time, when dealing with many objects, can be too slow for some animations.

**Reply by vsquared** [reply]
It's possible to retrieve the vertices of the original object, but it doesn't do much good because the runtime apparently won't allow them to be set_filled a second time.  End of story unless we can change the runtime.  Anybody know what perVertexStyles is?  It's set to 'false' in the code that I saw.

https://github.com/processing/processing4/blob/main/core/src/processing/core/PShape.java#L2604

```
    if (vertices != null && perVertexStyles) {
      for (int i = 0; i < vertexCount; i++) {
        setFill(i, fill);
      }
    }
  }
```

**Reply by hx2A** [reply]
The issue is that when a PShape object gets drawn to the screen with the `shape()` method, the vertex fill colors become permanent. You can change the fills as many times as you want to before calling the `shape()` method. After, the fills can't be updated again.

**Reply by villares** [reply]
There is a mention to an `update()` method for PShape at https://processing.github.io/processing-javadocs/core/processing/core/PShape.html, but I couldn't find it. 

If I have the time and energy I might investigate PShapeOpenGL https://processing.github.io/processing-javadocs/core/processing/opengl/PShapeOpenGL.html that can be instantiated from a PShape, it seems... I wonder if it would be an expensive operation to "copy" a pristine PShape into something I would then colorize.

**Reply by hx2A** [reply]
> If I have the time and energy I might investigate PShapeOpenGL https://processing.github.io/processing-javadocs/core/processing/opengl/PShapeOpenGL.html that can be instantiated from a PShape, it seems... I wonder if it would be an expensive operation to "copy" a pristine PShape into something I would then colorize.

PShapeOpenGL is an ugly mess. There is "pre-tessellated" geometry, which is the result of the Processing method calls, and is understandable and readable. This gets converted to tessellated geometry, and that's what actually gets sent to OpenGL to be rendered. That part is confusing and I understand very little of it, despite many efforts to figure it out.

It's the tessellated geometry that needs to get refreshed if you want `set_fills()` to change the colors after the shape is drawn to the screen (which is when tessellation happens).

**Reply by vsquared** [reply]
The Processing runtime code was changed: (https://github.com/processing/processing4/issues/677)
setFill(indx,color) used to work correctly, but no longer does.

**Reply by hx2A** [reply]
One of the Processing teams' priorities for 2025 is to make unit tests, which would help prevent this kind of bug from happening in the future.

**Reply by vsquared** [reply]
I resubmitted a bug report to go along with the one approximately a year ago.

The following may or may not be important.  It was committed 5 yr. ago and is over 5000 lines of code:
https://github.com/processing/processing/blob/master/core/src/processing/opengl/PShapeOpenGL.java

**Reply by vsquared** [reply]
My bug report is here: https://github.com/processing/processing4/issues/900#issuecomment-2574459691

This fix was just posted and it works!
```processing
PShape t;

void setup() {
  size(400, 400, P2D);
  surface.setTitle("Should change color when mouse clicked.");
  t = createShape();
  t.beginShape();
  t.fill(color(random(255), random(255), random(255)));
  t.vertex(200, 100);
  t.vertex(100, 300);
  t.vertex(300, 300);
  t.endShape(CLOSE);
  for (int i = 0; i < 3; i++) {
    t.setFill(i, color(random(255), random(255), random(255)));
  }
}

void draw() {
  background(209);
  shape(t);
}

void mousePressed() {
  for (int i = 0; i < t.getVertexCount(); i++) {
    ((PShapeOpenGL) t).beginTessellation(POINTS);
    t.setFill(i, color(random(255), random(255), random(255)));
    ((PShapeOpenGL) t).endTessellation();
  }  
}

```
**Addendum:**
mousePressed() may be shortened to this:
```processing
void mousePressed() {
  for (int i = 0; i < t.getVertexCount(); i++) {
    t.beginTessellation();
    t.setFill(i, color(random(255), random(255), random(255)));
    t.endTessellation();
  }  
}
```
Looks like you'll need to add _begin_tessellation()_ and _end_tessellation()_  to your py5 vocabulary to use the fix here.

**Reply by hx2A** [reply]
Excellent! Happy to see progress.

> Looks like you'll need to add begin_tessellation() and end_tessellation() to your py5 vocabulary to use the fix here.

I could to that but I think a better fix is to fix it in Processing, so Processing users can benefit from the insight.

**Reply by vsquared** [reply]
` I think a better fix is to fix it in Processing, so Processing users can benefit from the insight. `
Please encourage the author to submit a pull request after careful review to make sure it's the right thing to do.  It appears that the author knows what section of code to fix.

Proposed changes are to this code:
https://github.com/processing/processing4/blob/main/core/src/processing/opengl/PShapeOpenGL.java#L2007
```processing
--- a/core/src/processing/opengl/PShapeOpenGL.java
+++ b/core/src/processing/opengl/PShapeOpenGL.java
@@ -2005,12 +2005,12 @@ public class PShapeOpenGL extends PShape {
     }
 
     if (image == null) {
-      if (root.tessUpdate) {
+      inGeo.colors[index] = PGL.javaToNativeARGB(fill);
+      if (shapeCreated && tessellated && hasPolys) {
         int tessIdx = firstPolyVertex + index;
         tessGeo.polyColors[tessIdx] = PGL.javaToNativeARGB(fill);
         root.setModifiedPolyColors(tessIdx, tessIdx);
       } else {
-        inGeo.colors[index] = PGL.javaToNativeARGB(fill);
         markForTessellation();
       }
     }

```
**Addendum:**
One could argue that there is no 'bug'; the programmer just has to use pre-existing beginTessellation...endTessellation Processing calls to secondarily change vertex colors.
